File: 20-valid-parentheses/valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        // every time a closing bracket come 
        // check stack top is opening bracket or not 
        
        Map<Character, Character> map = new HashMap<>();
        map.put(')', '(');
        map.put(']', '[');
        map.put('}', '{');
        Stack<Character> st = new Stack<>();
        for(char c: s.toCharArray()){
            if(map.containsKey(c)){
                Character stackTop = st.isEmpty() ? '*' : st.peek();
                if(stackTop != map.get(c)) return false;
                st.pop();
            }else{
                st.push(c);
            }
        }
        return st.isEmpty();
    }
}

File: 20-valid-parentheses/valid-parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        unordered_map<char, char> map;// for each closing bracked , there should an opening bracket waiting to be closed.
        map[')'] = '(';
        map[']'] = '[';
        map['}'] = '{';
        stack<char> st; // opening bracket waiting to be closed 

        for(auto it: s){
            if(map.contains(it)){// closing bracket 
                int watingOpeningBrakcet = st.empty() ? '*' : st.top(); // handle stack empty case
                if(map[it] != watingOpeningBrakcet){
                    return false;// other closing bracket is waiting 
                }else{
                    st.pop();
                }
            }else{
                st.push(it); // wait for a closing bracket
            }
        }

        return st.empty(); // is any open bracket waiting to be open 
        
    }
};